controllers/core/cninode_controller.go

```diff
@@ -0,0 +1,101 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//     http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package controllers
+
+import (
+	"context"
+
+	"github.com/aws/amazon-vpc-resource-controller-k8s/apis/vpcresources/v1alpha1"
+	rcHealthz "github.com/aws/amazon-vpc-resource-controller-k8s/pkg/healthz"
+	"github.com/aws/amazon-vpc-resource-controller-k8s/pkg/k8s"
+	"github.com/aws/amazon-vpc-resource-controller-k8s/pkg/node/manager"
+	"github.com/go-logr/logr"
+
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/healthz"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+)
+
+type CNINodeController struct {
+	Log         logr.Logger
+	NodeManager manager.Manager
+	Context     context.Context
+	K8sAPI      k8s.K8sWrapper
+}
+
+func (s *CNINodeController) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	cniNode, err := s.K8sAPI.GetCNINode(req.Name, req.Namespace)
+	if err != nil {
+		s.Log.Error(err, "Failed to get CNINode")
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	if len(cniNode.Spec.Features) == 0 {
+		return ctrl.Result{}, nil
+	}
+
+	nodeName := cniNode.Name
+	s.Log.Info("The CNINode has been updated", "CNINode", nodeName, "Features", cniNode.Spec.Features)
+
+	if node, err := s.K8sAPI.GetNode(nodeName); err != nil {
+		// if not found the node, we don't requeue and just wait VPC CNI to send another request
+		s.Log.Info("CNINode reconciler didn't find the node and will do exponential retry", "Node", node.Name)
+		return ctrl.Result{}, err
+	} else {
+		// make sure we have the node in cache already
+		if _, found := s.NodeManager.GetNode(nodeName); !found {
+			s.Log.Info("CNINode controller could not find node, will try again", "NodeName", nodeName)
+			return ctrl.Result{Requeue: true}, nil
+		} else {
+			// add the node into working queue as UPDATE event
+			s.Log.Info("CNINode controller found the node and sending the node update")
+			if err = s.NodeManager.UpdateNode(nodeName); err != nil {
+				return ctrl.Result{}, err
+			}
+		}
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (s *CNINodeController) SetupWithManager(mgr ctrl.Manager, healthzHandler *rcHealthz.HealthzHandler) error {
+	healthzHandler.AddControllersHealthCheckers(
+		map[string]healthz.Checker{
+			"health-cninode-controller": s.Check(),
+		},
+	)
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&v1alpha1.CNINode{}).WithEventFilter(
+		eventFilters(),
```

Comment:
Why doesn't the default event filter work for us?

Reply:
I think the default filter is allowing all 4 ops, isn't it? I only want to watch for valid Update and not delete. I think even create dosen't make sense.

Reply:
I wouldn't worry about an extra reconcile loop on the creation. Ideally, controllers should just operate against observed state and not be aware of whether or not the object was created, updated, or refreshed (after a crash/restart/resync interval). In general, less deviations from the defaults == simpler and easier to reason about.

Reply:
Ack. I will remove the customied filter. Thanks